src/mscorlib/shared/System/IO/PinnedBufferMemoryStream.cs

```diff
@@ -30,18 +30,12 @@ internal PinnedBufferMemoryStream(byte[] array)
             Debug.Assert(array != null, "Array can't be null");
 
             int len = array.Length;
-            // Handle 0 length byte arrays specially.
-            if (len == 0)
-            {
-                array = new byte[1];
-                len = 0;
-            }
 
             _array = array;
             _pinningHandle = GCHandle.Alloc(array, GCHandleType.Pinned);
             // Now the byte[] is pinned for the lifetime of this instance.
             // But I also need to get a pointer to that block of memory...
-            fixed (byte* ptr = &_array[0])
+            fixed (byte* ptr = &MemoryMarshal.GetReference((Span<byte>)array))
                 Initialize(ptr, len, len, FileAccess.Read);
```

Comment:
Why is span needed here? fixed works fine with zero-length arrays).

Reply:
Pinning zero length array gives back `null`. `Initialize` method not accept null.

Reply:
I see.  Though that could be addressed as well just by passing `ptr != null ? ptr : (byte*)1` instead of `ptr`.

Reply:
Yes, there are several ways to address this. Another option in this case is to fix or delete the null check in the `Initialize` method.

- `(byte*)1` is a very unusual pointer value. If I saw it somewhere, I would be looking for memory corruption. It would be nice for the unmanaged pointers to either be null or point to valid memory.
- The code size varies for the different options. I suspect that `ptr != null ? ptr : (byte*)1` pattern is larger overall native code compared to `MemoryMarshal.GetReference` - we should measure.

Reply:
As now its
```asm
       call     GCHandle:InternalAlloc(ref,int):long
       or       rax, 1
       mov      qword ptr [rsi+96], rax
       lea      r8, bword ptr [rdi+16]
       mov      r9d, dword ptr [rdi+8]
       mov      bword ptr [rsp+28H], r8
       mov      rdx, r8
       mov      dword ptr [rsp+20H], 1
       movsxd   r8, ebx
       movsxd   r9, ebx
       mov      rcx, rsi
       call     UnmanagedMemoryStream:Initialize(long,long,long,int):this
```
Changing to
```csharp
fixed (byte* ptr = array)
    Initialize(ptr != null ? ptr : (byte*)1, len, len, FileAccess.Read);
```
It becomes
```asm
       call     GCHandle:InternalAlloc(ref,int):long
       or       rax, 1
       mov      qword ptr [rsi+96], rax
       mov      gword ptr [rsp+28H], rdi
       mov      r8, gword ptr [rsp+28H]
       cmp      dword ptr [r8+8], 0
       jne      SHORT G_M3256_IG04
G_M3256_IG03:
       xor      r8, r8
       jmp      SHORT G_M3256_IG05
G_M3256_IG04:
       mov      r8, gword ptr [rsp+28H]
       cmp      dword ptr [r8+8], 0
       jbe      SHORT G_M3256_IG09
       mov      r8, gword ptr [rsp+28H]
       add      r8, 16
G_M3256_IG05:
       test     r8, r8
       jne      SHORT G_M3256_IG06
       mov      edx, 1
       jmp      SHORT G_M3256_IG07
G_M3256_IG06:
       mov      rdx, r8
G_M3256_IG07:
       mov      dword ptr [rsp+20H], 1
       movsxd   r9, ebx
       mov      r8, r9
       mov      rcx, rsi
       call     UnmanagedMemoryStream:Initialize(long,long,long,int):this
```
Which seems more involved?